src/storage/conversion/ChainedConverter.ts

```diff
@@ -1,89 +1,305 @@
+        bestMatch = { type: leftType, weight };
+        if (weight === 1) {
+          return bestMatch;
+        }
+      }
+    }
+  }
+
+  return bestMatch;
+}
+
+/**
+ * An LRU cache for storing `ConversionPath`s.
+ */
+class LruPathCache {
+  private readonly maxSize: number;
+  private readonly paths: ConversionPath[] = [];
+
+  public constructor(maxSize: number) {
+    this.maxSize = maxSize;
+  }
+
+  /**
+   * Add the given path to the cache as most recently used.
+   */
+  public add(path: ConversionPath): void {
+    this.paths.push(path);
+    if (this.paths.length > this.maxSize) {
+      this.paths.shift();
+    }
+  }
+
+  /**
+   * Find a path that can convert the given type to the given preferences.
+   * Note that this finds the first matching path in the cache,
+   * not the best one, should there be multiple results.
+   * In practice this should almost never be the case though.
+   */
+  public find(inType: string, outPreferences: ValuePreferences): MatchedPath | undefined {
+    // Last element has more chance of being the correct one
+    for (let i = this.paths.length - 1; i >= 0; --i) {
+      const path = this.paths[i];
+      const match = this.getMatchedPath(inType, outPreferences, path);
+      if (match.weight) {
+        // Set matched path to most recent result in the cache
+        this.paths.splice(i, 1);
+        this.paths.push(path);
+        return match;
+      }
+    }
+  }
+
+  /**
+   * Calculates the weights and exact types when using the given path on the given type and preferences.
+   * Weight will be 0 if there is no match.
+   */
+  private getMatchedPath(inType: string, outPreferences: ValuePreferences, path: ConversionPath): MatchedPath {
+    const inWeight = getMatchingType({ [inType]: 1 }, path.inTypes).weight;
+    if (inWeight === 0) {
+      return { path, inType, outType: 'unknown', weight: 0 };
+    }
+    const outResult = getMatchingType(outPreferences, path.outTypes);
+    return { path, inType, outType: outResult.type, weight: inWeight * outResult.weight };
+  }
+}
 
 /**
  * A meta converter that takes an array of other converters as input.
- * It chains these converters by finding intermediate types that are supported by converters on either side.
+ * It chains these converters by finding a path of converters
+ * that can go from the given content-type to the given type preferences.
+ * In case there are multiple paths, the shortest one with the highest weight gets found.
+ * Will error in case no path can be found.
+ *
+ * Generated paths get stored in an internal cache for later re-use on similar requests.
+ * Note that due to this caching `RepresentationConverter`s
+ * that change supported input/output types at runtime are not supported,
+ * unless cache size is set to 0.
+ *
+ * This is not a TypedRepresentationConverter since the supported output types
+ * might depend on what is the input content-type.
```

Comment:
Is that a problem? We could just list all possible types; no guarantee that any concrete conversion exists anyway.

Reply:
Wonder if this is a cause for the asymmetry.

Reply:
Listing all types is also a possibility. It's just that I didn't see any advantage in making it a typed converter.

Reply:
At least you know then what types it definitely does _not_ generate.

Reply:
It sort of already has that since no valid path would be found. We could also check the output types immediately but that's one of the suggestions in the possible optimizations list.

The reason I don't want to do it is because it sort of breaks the contract of a typed converter where it is expected that one can go from all input types to all output types.

Why exactly would you want to do this, is it to have better logging/error messages about (possible) supported types? Because otherwise those can be added on their own.

Reply:
> where it is expected that one can go from all input types to all output types.

Is that so? I thought it was a suggestion.

> Why exactly would you want to do this

When convertors start being used in more complex situations, to quickly find convertors that could be relevant.

But no strong opinion, so ok with leaving as-is.